craft_application/manifest.py

```diff
@@ -0,0 +1,96 @@
+# This file is part of craft-application.
+#
+# Copyright 2024 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify it
+# under the terms of the GNU Lesser General Public License version 3, as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+# SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Functions to generate the craft manifest of an artifact."""
+
+import hashlib
+import pathlib
+from datetime import datetime, timezone
+from typing import Any
+
+from craft_application.models import BuildInfo, Project
+from craft_application.models.manifest import (
+    CraftManifest,
+    Hashes,
+    ProjectManifest,
+    SessionArtifactManifest,
+)
+
+
+def from_packed_artifact(
+    project: Project, build_info: BuildInfo, artifact: pathlib.Path
+) -> ProjectManifest:
+    """Create the project manifest for a packed artifact."""
+    hashes = _compute_hashes(artifact)
+
+    now = datetime.now(timezone.utc)
+
+    return ProjectManifest.unmarshal(
+        {
+            "component-name": project.name,
+            "component-version": project.version,
+            "component-description": project.summary,
+            "component-id": {"hashes": hashes.marshal()},
+            "architecture": build_info.build_for,
+            "license": project.license,
+            "creation_timestamp": now.isoformat(),
+        }
+    )
+
+
+def from_session_report(report: dict[str, Any]) -> list[SessionArtifactManifest]:
+    """Create session manifests from a fetch-session report."""
+    artifacts: list[SessionArtifactManifest] = []
+    for artifact in report["artefacts"]:
+        metadata = artifact["metadata"]
+        data = {
+            "type": metadata["type"],
+            "component-name": metadata["name"],
+            "component-version": metadata["version"],
+            "component-description": metadata["description"],
+            "architecture": "TODO",
```

Comment:
@cmatsuoka should we add the architecture to the fetch-service artifact metadata?

Reply:
Yes, this is easy to obtain and manifest users will likely want this information despite the fact that it's not in the minimal sbom requirements.

Reply:
Added an issue on fetch-service

Reply:
You can assign the value from the `architecture` field if it exists. The Fetch Service will populate it when applicable.